File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.cs
public class Solution {
    public int[] TopKFrequent(int[] nums, int k) {
        // I would use a HashMap to store each num frequence. 
        // Also initialise a priority queue that queues each num in the hasmap based on their frequency 
        // then convert and return the priorty queue as an array of int
        // Time Complexity would be O(nlogn) where n is the number of nums 
        // Space Complexity would be O(n^2) where n is the length of nums

        var hashMap = new Dictionary<int, int>();
        var pq = new PriorityQueue<int, int>();

        foreach (var n in nums) {
            if(!hashMap.ContainsKey(n)) hashMap[n] = 0;
            hashMap[n]++;
        } 

        foreach (var kvp in hashMap) {
            pq.Enqueue(kvp.Key, kvp.Value);

            if(pq.Count > k) {
                var removed = pq.Dequeue();
                hashMap.Remove(removed);
            }
        }

        return hashMap.Keys.ToArray();
    }
}

File: Data Structures & Algorithms/kth-largest-element-in-an-array/submission-2.js


class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number}
     */
    findKthLargest(nums, k) {
        const pq = new MinPriorityQueue();

        // queue nums 
        // O(nlogk)
        for (const n of nums) {
            pq.enqueue(n)

            if (pq.size() > k) pq.dequeue();
        }

        // dequque kth largest
        return pq.dequeue();
    }
}
